File: js/health_vault/alert_engine.js
/**
 * HC-301 — Browser Alert Engine (mirrors backend HCAlertEngine / AlertEngine).
 * Observational only — not a diagnosis. Local-first; never sends off-device.
 */
(function (global) {
  "use strict";

  const SAFETY_DISCLAIMER =
    "Observational only — not a diagnosis. HealthChecker+ does not replace " +
    "FreeStyle Libre alarms, Samsung Health Monitor, medical care, or emergency services. " +
    "No medication or insulin dosing advice is provided.";

  const SEVERITIES = ["informational", "watch", "warning", "urgent", "critical"];
  const SEVERITY_RANK = {};
  SEVERITIES.forEach((s, i) => {
    SEVERITY_RANK[s] = i;
  });

  const LS_KEY = "HC_GUARDIAN_ALERTS_V1";
  const DEFAULT_COOLDOWNS = {
    informational: 360,
    watch: 180,
    warning: 60,
    urgent: 30,
    critical: 15,
  };

  function utcNow() {
    return new Date().toISOString().replace(/\.\d{3}Z$/, "Z");
  }

  function uuid() {
    if (global.crypto && crypto.randomUUID) return crypto.randomUUID();
    return "a-" + Date.now().toString(36) + "-" + Math.random().toString(36).slice(2, 10);
  }

  function parseIso(ts) {
    if (!ts) return null;
    const t = Date.parse(String(ts).replace("Z", "+00:00"));
    return Number.isFinite(t) ? t / 1000 : null;
  }

  function addMinutes(isoTs, minutes) {
    const base = parseIso(isoTs);
    const ms = (base != null ? base * 1000 : Date.now()) + minutes * 60 * 1000;
    return new Date(ms).toISOString().replace(/\.\d{3}Z$/, "Z");
  }

  function loadAlertsRaw() {
    const Vault = global.HCHealthVault;
    if (Vault && typeof Vault.listAlerts === "function") {
      try {
        return Vault.listAlerts();
      } catch (_) {}
    }
    try {
      const raw = localStorage.getItem(LS_KEY);
      if (!raw) return [];
      const parsed = JSON.parse(raw);
      return Array.isArray(parsed) ? parsed : [];
    } catch (_) {
      return [];
    }
  }

  function saveAlertsRaw(alerts) {
    const Vault = global.HCHealthVault;
    if (Vault && typeof Vault.saveAlerts === "function") {
      try {
        Vault.saveAlerts(alerts);
        return;
      } catch (_) {}
    }
    try {
      localStorage.setItem(LS_KEY, JSON.stringify(alerts));
    } catch (_) {}
  }

  function upsertAlert(alert) {
    const Vault = global.HCHealthVault;
    if (Vault && typeof Vault.upsertAlert === "function") {
      try {
        return Vault.upsertAlert(alert);
      } catch (_) {}
    }
    const items = loadAlertsRaw();
    const id = alert.alert_id;
    let found = false;
    for (let i = 0; i < items.length; i++) {
      if (items[i].alert_id === id) {
        items[i] = alert;
        found = true;
        break;
      }
    }
    if (!found) items.push(alert);
    saveAlertsRaw(items);
    return alert;
  }

  function appendAudit(alert, at, action, detail) {
    const hist = Array.isArray(alert.audit_history) ? alert.audit_history.slice() : [];
    hist.push({ at: at, action: action, detail: detail || {} });
    alert.audit_history = hist;
  }

  function emit(name, payload) {
    const Bus = global.HCEventBus;
    if (!Bus || !Bus.publish) return;
    try {
      Bus.publish(name, {
        alert_id: payload && payload.alert_id,
        severity: payload && payload.severity,
      });
    } catch (_) {}
  }

  function HCAlertEngine(opts) {
    opts = opts || {};
    this.cooldowns = opts.cooldowns_minutes || Object.assign({}, DEFAULT_COOLDOWNS);
  }

  HCAlertEngine.SAFETY_DISCLAIMER = SAFETY_DISCLAIMER;
  HCAlertEngine.SEVERITIES = SEVERITIES;
  HCAlertEngine.SEVERITY_RANK = SEVERITY_RANK;

  HCAlertEngine.prototype.listAlerts = function (filter) {
    filter = filter || {};
    let items = loadAlertsRaw().slice();
    if (filter.patient_id) {
      items = items.filter((a) => a.patient_id === filter.patient_id);
    }
    if (filter.status) {
      items = items.filter((a) => a.status === filter.status);
    }
    if (filter.active_only) {
      items = items.filter((a) =>
        ["active", "acknowledged", "snoozed"].indexOf(a.status) >= 0
      );
    }
    return items;
  };

  HCAlertEngine.prototype.getAlert = function (alertId) {
    const items = loadAlertsRaw();
    for (let i = 0; i < items.length; i++) {
      if (items[i].alert_id === alertId) return items[i];
    }
    return null;
  };

  HCAlertEngine.prototype.ingestEvaluation = function (evaluation, opts) {
    opts = opts || {};
    if (!evaluation || !evaluation.triggered) return null;
    const nowTs = opts.now || utcNow();
    const patientId = opts.patient_id || "default-patient";
    let severity = String(evaluation.severity || "watch");
    if (!(severity in SEVERITY_RANK)) severity = "watch";
    const ruleId = String(evaluation.rule_id || "unknown");
    const metrics = (evaluation.metrics || []).slice();
    if (evaluation.metric && metrics.indexOf(evaluation.metric) < 0) {
      metrics.unshift(evaluation.metric);
    }
    const dedupe = String(
      evaluation.deduplication_key ||
        patientId + "|" + ruleId + "|" + metrics.slice().sort().join("+")
    );

    const existing = this._findOpenByDedupe(dedupe, patientId);
    if (existing) {
      return this._updateExisting(existing, evaluation, severity, opts.source_event_ids, nowTs);
    }

    if (this._inCooldown(dedupe, patientId, nowTs) && severity !== "critical") {
      return null;
    }

    const alert = {
      alert_id: uuid(),
      patient_id: patientId,
      rule_id: ruleId,
      rule_version: String(evaluation.rule_version || "1.0.0"),
      title: String(evaluation.title || ruleId),
      message: String(evaluation.message || evaluation.title || ruleId),
      severity: severity,
      category: String(evaluation.category || "general"),
      metrics: metrics,
      source_event_ids: (opts.source_event_ids || evaluation.source_event_ids || []).slice(),
      created_at: nowTs,
      updated_at: nowTs,
      first_detected_at: nowTs,
      last_detected_at: nowTs,
      occurrence_count: 1,
      status: "active",
      acknowledgement_state: "none",
      acknowledged_at: null,
      resolved_at: null,
      snoozed_until: null,
      cooldown_until: null,
      deduplication_key: dedupe,
      evidence: Object.assign({}, evaluation.evidence || {}),
      recommended_next_step: String(evaluation.recommended_next_step || ""),
      safety_disclaimer: String(evaluation.safety_disclaimer || SAFETY_DISCLAIMER),
      audit_history: [
        { at: nowTs, action: "created", detail: { severity: severity, rule_id: ruleId } },
      ],
    };
    const saved = upsertAlert(alert);
    emit("AlertCreated", saved);
    return saved;
  };

  HCAlertEngine.prototype.acknowledge = function (alertId, opts) {
    opts = opts || {};
    const alert = this.getAlert(alertId);
    if (!alert) return { ok: false, errors: ["alert_not_found"] };
    const nowTs = opts.now || utcNow();
    alert.acknowledgement_state = "acknowledged";
    alert.acknowledged_at = nowTs;
    alert.updated_at = nowTs;
    if (alert.status === "active") alert.status = "acknowledged";
    appendAudit(alert, nowTs, "acknowledged", { note: opts.note || null });
    const saved = upsertAlert(alert);
    emit("AlertAcknowledged", saved);
    return { ok: true, alert: saved };
  };

  HCAlertEngine.prototype.resolve = function (alertId, opts) {
    opts = opts || {};
    const alert = this.getAlert(alertId);
    if (!alert) return { ok: false, errors: ["alert_not_found"] };
    const nowTs = opts.now || utcNow();
    if (
      alert.severity === "critical" &&
      alert.acknowledgement_state !== "acknowledged" &&
      !opts.force
    ) {
      return { ok: false, errors: ["critical_requires_acknowledgement"], alert: alert };
    }
    alert.status = "resolved";
    alert.resolved_at = nowTs;
    alert.updated_at = nowTs;
    const cd = Number(this.cooldowns[String(alert.severity)] || 30);
    alert.cooldown_until = addMinutes(nowTs, cd);
    appendAudit(alert, nowTs, "resolved", { note: opts.note || null, force: !!opts.force });
    const saved = upsertAlert(alert);
    emit("AlertResolved", saved);
    return { ok: true, alert: saved };
  };

  HCAlertEngine.prototype.snooze = function (alertId, opts) {
    opts = opts || {};
    const alert = this.getAlert(alertId);
    if (!alert) return { ok: false, errors: ["alert_not_found"] };
    if (alert.severity === "critical") {
      return { ok: false, errors: ["critical_cannot_snooze"], alert: alert };
    }
    const nowTs = opts.now || utcNow();
    const minutes = opts.minutes != null ? opts.minutes : 60;
    alert.status = "snoozed";
    alert.snoozed_until = addMinutes(nowTs, minutes);
    alert.updated_at = nowTs;
    appendAudit(alert, nowTs, "snoozed", { minutes: minutes });
    const saved = upsertAlert(alert);
    emit("AlertSnoozed", saved);
    return { ok: true, alert: saved };
  };

  HCAlertEngine.prototype.activeCounts = function (patientId) {
    const counts = {};
    SEVERITIES.forEach((s) => {
      counts[s] = 0;
    });
    this.listAlerts({ patient_id: patientId || undefined, active_only: true }).forEach((a) => {
      if (a.severity in counts) counts[a.severity] += 1;
    });
    counts.total = SEVERITIES.reduce((n, s) => n + counts[s], 0);
    return counts;
  };

  HCAlertEngine.prototype._findOpenByDedupe = function (dedupe, patientId) {
    const items = this.listAlerts({ patient_id: patientId, active_only: true });
    for (let i = 0; i < items.length; i++) {
      if (items[i].deduplication_key === dedupe) return items[i];
    }
    return null;
  };

  HCAlertEngine.prototype._inCooldown = function (dedupe, patientId, nowTs) {
    const nowEpoch = parseIso(nowTs) || 0;
    const items = loadAlertsRaw();
    for (let i = 0; i < items.length; i++) {
      const a = items[i];
      if (a.patient_id !== patientId) continue;
      if (a.deduplication_key !== dedupe) continue;
      const until = parseIso(a.cooldown_until);
      if (until != null && until > nowEpoch) return true;
    }
    return false;
  };

  HCAlertEngine.prototype._updateExisting = function (
    existing,
    evaluation,
    severity,
    sourceEventIds,
    nowTs
  ) {
    const alert = Object.assign({}, existing);
    const prev = String(alert.severity || "watch");
    const escalated = (SEVERITY_RANK[severity] || 0) > (SEVERITY_RANK[prev] || 0);
    alert.occurrence_count = (Number(alert.occurrence_count) || 1) + 1;
    alert.last_detected_at = nowTs;
    alert.updated_at = nowTs;
    alert.evidence = Object.assign({}, evaluation.evidence || alert.evidence || {});
    if (sourceEventIds && sourceEventIds.length) {
      const ids = (alert.source_event_ids || []).slice();
      sourceEventIds.forEach((sid) => {
        if (ids.indexOf(sid) < 0) ids.push(sid);
      });
      alert.source_event_ids = ids;
    }
    if (escalated) {
      alert.severity = severity;
      if (alert.status === "acknowledged" || alert.status === "snoozed") {
        alert.status = "active";
        alert.acknowledgement_state = "none";
      }
      appendAudit(alert, nowTs, "escalated", {
        from: prev,
        to: severity,
        occurrence_count: alert.occurrence_count,
      });
      const saved = upsertAlert(alert);
      emit("AlertEscalated", saved);
      return saved;
    }
    appendAudit(alert, nowTs, "redetected", {
      occurrence_count: alert.occurrence_count,
      severity: alert.severity,
    });
    if (
      alert.occurrence_count >= 5 &&
      (prev === "watch" || prev === "warning" || prev === "informational")
    ) {
      const nextSev = SEVERITIES[Math.min((SEVERITY_RANK[prev] || 0) + 1, SEVERITIES.length - 1)];
      alert.severity = nextSev;
      alert.status = "active";
      appendAudit(alert, nowTs, "escalated_persistence", { from: prev, to: nextSev });
      const saved = upsertAlert(alert);
      emit("AlertEscalated", saved);
      return saved;
    }
    const saved = upsertAlert(alert);
    emit("AlertUpdated", saved);
    return saved;
  };

  global.HCAlertEngine = HCAlertEngine;
})(typeof window !== "undefined" ? window : globalThis);

File: backend/health_vault/guardian/alert_engine.py
"""HC-301 formal Alert Engine — observational, auditable, non-diagnostic."""

from __future__ import annotations

from dataclasses import asdict, dataclass, field
from typing import Any
from uuid import uuid4

from backend.health_vault.models import utc_now

SAFETY_DISCLAIMER = (
    "Observational only — not a diagnosis. HealthChecker+ does not replace "
    "FreeStyle Libre alarms, Samsung Health Monitor, medical care, or emergency services. "
    "No medication or insulin dosing advice is provided."
)

SEVERITIES = ("informational", "watch", "warning", "urgent", "critical")
SEVERITY_RANK = {s: i for i, s in enumerate(SEVERITIES)}
STATUSES = ("active", "acknowledged", "snoozed", "resolved", "expired")


@dataclass
class Alert:
    alert_id: str = field(default_factory=lambda: str(uuid4()))
    patient_id: str = "default-patient"
    rule_id: str = ""
    rule_version: str = "1.0.0"
    title: str = ""
    message: str = ""
    severity: str = "watch"
    category: str = "general"
    metrics: list[str] = field(default_factory=list)
    source_event_ids: list[str] = field(default_factory=list)
    created_at: str = field(default_factory=utc_now)
    updated_at: str = field(default_factory=utc_now)
    first_detected_at: str = field(default_factory=utc_now)
    last_detected_at: str = field(default_factory=utc_now)
    occurrence_count: int = 1
    status: str = "active"
    acknowledgement_state: str = "none"
    acknowledged_at: str | None = None
    resolved_at: str | None = None
    snoozed_until: str | None = None
    cooldown_until: str | None = None
    deduplication_key: str = ""
    evidence: dict[str, Any] = field(default_factory=dict)
    recommended_next_step: str = ""
    safety_disclaimer: str = SAFETY_DISCLAIMER
    audit_history: list[dict[str, Any]] = field(default_factory=list)

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "Alert":
        known = {f.name for f in cls.__dataclass_fields__.values()}  # type: ignore[attr-defined]
        payload = {k: v for k, v in (data or {}).items() if k in known}
        if "metrics" in payload and payload["metrics"] is None:
            payload["metrics"] = []
        if "source_event_ids" in payload and payload["source_event_ids"] is None:
            payload["source_event_ids"] = []
        if "audit_history" in payload and payload["audit_history"] is None:
            payload["audit_history"] = []
        if "evidence" in payload and payload["evidence"] is None:
            payload["evidence"] = {}
        return cls(**payload)


def _parse_iso(ts: str | None) -> float | None:
    if not ts:
        return None
    try:
        from datetime import datetime

        return datetime.fromisoformat(ts.replace("Z", "+00:00")).timestamp()
    except Exception:
        return None


def _add_minutes(iso_ts: str, minutes: float) -> str:
    from datetime import datetime, timedelta, timezone

    try:
        dt = datetime.fromisoformat(iso_ts.replace("Z", "+00:00"))
    except Exception:
        dt = datetime.now(timezone.utc)
    return (dt + timedelta(minutes=minutes)).replace(microsecond=0).isoformat().replace("+00:00", "Z")


class AlertEngine:
    """Create/update/acknowledge/resolve alerts with dedupe, cooldown, escalation."""

    def __init__(
        self,
        store: Any,
        *,
        cooldowns_minutes: dict[str, int] | None = None,
        bus: Any | None = None,
    ) -> None:
        self.store = store
        self.cooldowns = cooldowns_minutes or {
            "informational": 360,
            "watch": 180,
            "warning": 60,
            "urgent": 30,
            "critical": 15,
        }
        self.bus = bus

    def list_alerts(
        self,
        *,
        patient_id: str | None = None,
        status: str | None = None,
        active_only: bool = False,
    ) -> list[dict[str, Any]]:
        items = list(self.store.list_alerts())
        if patient_id:
            items = [a for a in items if a.get("patient_id") == patient_id]
        if status:
            items = [a for a in items if a.get("status") == status]
        if active_only:
            items = [a for a in items if a.get("status") in ("active", "acknowledged", "snoozed")]
        return items

    def get_alert(self, alert_id: str) -> dict[str, Any] | None:
        for a in self.store.list_alerts():
            if a.get("alert_id") == alert_id:
                return a
        return None

    def ingest_evaluation(
        self,
        evaluation: dict[str, Any],
        *,
        patient_id: str = "default-patient",
        source_event_ids: list[str] | None = None,
        now: str | None = None,
    ) -> dict[str, Any] | None:
        """Upsert an alert from a structured rule evaluation result."""
        if not evaluation or not evaluation.get("triggered"):
            return None
        if not isinstance(evaluation, dict):
            return None
        if not evaluation.get("rule_id"):
            return None
        severity = str(evaluation.get("severity") or "watch")
        if severity not in SEVERITY_RANK:
            return None
        now_ts = now or utc_now()
        patient_id = str(patient_id or "default-patient")
        rule_id = str(evaluation.get("rule_id") or "unknown")
        metrics = list(evaluation.get("metrics") or [])
        if evaluation.get("metric") and evaluation["metric"] not in metrics:
            metrics.insert(0, evaluation["metric"])
        dedupe = str(
            evaluation.get("deduplication_key")
            or f"{patient_id}|{rule_id}|{'+'.join(sorted(str(x) for x in metrics))}"
        )
        # Enforce patient prefix on dedupe keys
        if not dedupe.startswith(f"{patient_id}|"):
            dedupe = f"{patient_id}|{dedupe}"

        existing = self._find_open_by_dedupe(dedupe, patient_id)
        if existing:
            return self._update_existing(existing, evaluation, severity, source_event_ids, now_ts)

        # Cooldown: suppress brand-new alert if a recently resolved one shares dedupe
        if self._in_cooldown(dedupe, patient_id, now_ts) and severity != "critical":
            return None

        alert = Alert(
            patient_id=patient_id,
            rule_id=rule_id,
            rule_version=str(evaluation.get("rule_version") or "1.0.0"),
            title=str(evaluation.get("title") or rule_id),
            message=str(evaluation.get("message") or evaluation.get("title") or rule_id),
            severity=severity,
            category=str(evaluation.get("category") or "general"),
            metrics=metrics,
            source_event_ids=list(source_event_ids or evaluation.get("source_event_ids") or []),
            created_at=now_ts,
            updated_at=now_ts,
            first_detected_at=now_ts,
            last_detected_at=now_ts,
            deduplication_key=dedupe,
            evidence=dict(evaluation.get("evidence") or {}),
            recommended_next_step=str(evaluation.get("recommended_next_step") or ""),
            safety_disclaimer=str(evaluation.get("safety_disclaimer") or SAFETY_DISCLAIMER),
            audit_history=[
                {
                    "at": now_ts,
                    "action": "created",
                    "detail": {"severity": severity, "rule_id": rule_id},
                }
            ],
        )
        saved = self.store.upsert_alert(alert.to_dict())
        self._emit("AlertCreated", saved)
        return saved

    def acknowledge(
        self,
        alert_id: str,
        *,
        note: str | None = None,
        now: str | None = None,
    ) -> dict[str, Any]:
        alert = self.get_alert(alert_id)
        if not alert:
            return {"ok": False, "errors": ["alert_not_found"]}
        now_ts = now or utc_now()
        # Critical alerts stay visible (status acknowledged, not dismissed)
        alert["acknowledgement_state"] = "acknowledged"
        alert["acknowledged_at"] = now_ts
        alert["updated_at"] = now_ts
        if alert.get("status") == "active":
            alert["status"] = "acknowledged"
        self._append_audit(alert, now_ts, "acknowledged", {"note": note})
        saved = self.store.upsert_alert(alert)
        self._emit("AlertAcknowledged", saved)
        return {"ok": True, "alert": saved}

    def resolve(
        self,
        alert_id: str,
        *,
        note: str | None = None,
        force: bool = False,
        now: str | None = None,
    ) -> dict[str, Any]:
        alert = self.get_alert(alert_id)
        if not alert:
            return {"ok": False, "errors": ["alert_not_found"]}
        now_ts = now or utc_now()
        severity = alert.get("severity")
        # Critical cannot be silently dismissed — must be acknowledged first unless force (admin/test)
        if severity == "critical" and alert.get("acknowledgement_state") != "acknowledged" and not force:
            return {
                "ok": False,
                "errors": ["critical_requires_acknowledgement"],
                "alert": alert,
            }
        alert["status"] = "resolved"
        alert["resolved_at"] = now_ts
        alert["updated_at"] = now_ts
        cd = int(self.cooldowns.get(str(severity), 30))
        alert["cooldown_until"] = _add_minutes(now_ts, cd)
        self._append_audit(alert, now_ts, "resolved", {"note": note, "force": force})
        saved = self.store.upsert_alert(alert)
        self._emit("AlertResolved", saved)
        return {"ok": True, "alert": saved}

    def snooze(
        self,
        alert_id: str,
        *,
        minutes: int = 60,
        now: str | None = None,
    ) -> dict[str, Any]:
        alert = self.get_alert(alert_id)
        if not alert:
            return {"ok": False, "errors": ["alert_not_found"]}
        if alert.get("severity") == "critical":
            return {"ok": False, "errors": ["critical_cannot_snooze"], "alert": alert}
        now_ts = now or utc_now()
        alert["status"] = "snoozed"
        alert["snoozed_until"] = _add_minutes(now_ts, minutes)
        alert["updated_at"] = now_ts
        self._append_audit(alert, now_ts, "snoozed", {"minutes": minutes})
        saved = self.store.upsert_alert(alert)
        self._emit("AlertSnoozed", saved)
        return {"ok": True, "alert": saved}

    def active_counts(self, patient_id: str | None = None) -> dict[str, int]:
        counts = {s: 0 for s in SEVERITIES}
        for a in self.list_alerts(patient_id=patient_id, active_only=True):
            sev = a.get("severity")
            if sev in counts:
                counts[sev] += 1
        counts["total"] = sum(counts[s] for s in SEVERITIES)
        return counts

    def _find_open_by_dedupe(self, dedupe: str, patient_id: str) -> dict[str, Any] | None:
        for a in self.list_alerts(patient_id=patient_id, active_only=True):
            if a.get("deduplication_key") == dedupe:
                return a
        return None

    def _in_cooldown(self, dedupe: str, patient_id: str, now_ts: str) -> bool:
        now_epoch = _parse_iso(now_ts) or 0.0
        for a in self.store.list_alerts():
            if a.get("patient_id") != patient_id:
                continue
            if a.get("deduplication_key") != dedupe:
                continue
            until = _parse_iso(a.get("cooldown_until"))
            if until is not None and until > now_epoch:
                return True
        return False

    def _update_existing(
        self,
        existing: dict[str, Any],
        evaluation: dict[str, Any],
        severity: str,
        source_event_ids: list[str] | None,
        now_ts: str,
    ) -> dict[str, Any]:
        alert = dict(existing)
        prev = str(alert.get("severity") or "watch")
        escalated = SEVERITY_RANK.get(severity, 0) > SEVERITY_RANK.get(prev, 0)
        alert["occurrence_count"] = int(alert.get("occurrence_count") or 1) + 1
        alert["last_detected_at"] = now_ts
        alert["updated_at"] = now_ts
        alert["evidence"] = dict(evaluation.get("evidence") or alert.get("evidence") or {})
        if source_event_ids:
            ids = list(alert.get("source_event_ids") or [])
            for sid in source_event_ids:
                if sid not in ids:
                    ids.append(sid)
            alert["source_event_ids"] = ids
        if escalated:
            alert["severity"] = severity
            # Re-open visibility for escalations
            if alert.get("status") in ("acknowledged", "snoozed"):
                alert["status"] = "active"
                alert["acknowledgement_state"] = "none"
            self._append_audit(
                alert,
                now_ts,
                "escalated",
                {"from": prev, "to": severity, "occurrence_count": alert["occurrence_count"]},
            )
            saved = self.store.upsert_alert(alert)
            self._emit("AlertEscalated", saved)
            return saved
        self._append_audit(
            alert,
            now_ts,
            "redetected",
            {"occurrence_count": alert["occurrence_count"], "severity": alert.get("severity")},
        )
        # Persistence escalation: same severity but many repeats → bump one level
        if int(alert["occurrence_count"]) >= 5 and prev in ("watch", "warning", "informational"):
            next_sev = SEVERITIES[min(SEVERITY_RANK[prev] + 1, len(SEVERITIES) - 1)]
            alert["severity"] = next_sev
            alert["status"] = "active"
            self._append_audit(
                alert,
                now_ts,
                "escalated_persistence",
                {"from": prev, "to": next_sev},
            )
            saved = self.store.upsert_alert(alert)
            self._emit("AlertEscalated", saved)
            return saved
        saved = self.store.upsert_alert(alert)
        self._emit("AlertUpdated", saved)
        return saved

    def _append_audit(self, alert: dict[str, Any], at: str, action: str, detail: dict[str, Any]) -> None:
        hist = list(alert.get("audit_history") or [])
        hist.append({"at": at, "action": action, "detail": detail})
        alert["audit_history"] = hist

    def _emit(self, name: str, payload: dict[str, Any]) -> None:
        if self.bus is not None:
            try:
                self.bus.publish(name, {"alert_id": payload.get("alert_id"), "severity": payload.get("severity")})
            except Exception:
                pass
